### PR TITLE
Wii classic controller and nunchuk support

### PR DESCRIPTION
**Pull request by Anh302kay**
Adds Wii classic controller support. Buttons are inverted to be correct with the GUI. Analog sticks are supported now, deadzone might need adjusting. Tested on real hardware with snes classic controller. Also now supports the wii nunchuk, tested on real hardware.

### CODE REVIEW

**Comment by Mefiresu** [comment]
Thanks ! Though I believe the best way to implement the inverted buttons would be to use the custom settings available here: https://github.com/Rubberduckycooly/RSDKv5-Decompilation/blob/master/RSDKv5/RSDK/User/Core/UserCore.cpp#L319

It might be necessary to flip the Wiimote inputs too once these options are set.
Could you research this ?

File: RSDKv5/RSDK/Input/Wii/WiiInputDevice.cpp [hunk 1]
```diff
@@ -6,20 +6,61 @@ void RSDK::SKU::InputDeviceWii::UpdateInput() {
...
-    this->stateA      = (this->buttonMasks & WPAD_BUTTON_2) != 0;
-    this->stateB      = (this->buttonMasks & WPAD_BUTTON_1) != 0;
-    this->stateC      = (this->buttonMasks & 0) != 0;
-    this->stateX      = (this->buttonMasks & WPAD_BUTTON_A) != 0;
-    this->stateY      = (this->buttonMasks & WPAD_BUTTON_B) != 0;
-    this->stateZ      = (this->buttonMasks & 0) != 0;
-    this->stateStart  = (this->buttonMasks & WPAD_BUTTON_PLUS) != 0;
-    this->stateSelect = (this->buttonMasks & WPAD_BUTTON_MINUS) != 0;
-
+    int type = data->exp.type;
```
Comment (Mefiresu) [inline RSDKv5/RSDK/Input/Wii/WiiInputDevice.cpp:9]: Missing `WPAD_Data()` ?
```suggestion
    WPADData *data = WPAD_Data(0);
    int type = data->exp.type;
```

File: RSDKv5/RSDK/Input/Wii/WiiInputDevice.cpp [hunk 2]
```diff
@@ -6,20 +6,61 @@ void RSDK::SKU::InputDeviceWii::UpdateInput() {
...
+            this->stateZ        = (this->buttonMasks & 0) != 0;
+            this->stateStart    = (this->buttonMasks & WPAD_CLASSIC_BUTTON_PLUS) != 0;
+            this->stateSelect   = (this->buttonMasks & WPAD_CLASSIC_BUTTON_MINUS) != 0;
+            this->stateUp       |= (data->exp.classic.ljs.pos.y > data->exp.classic.ljs.center.y + 5) ? 1 : 0;
+            this->stateDown     |= (data->exp.classic.ljs.pos.y < data->exp.classic.ljs.center.y - 5) ? 1 : 0;
+            this->stateLeft     |= (data->exp.classic.ljs.pos.x < data->exp.classic.ljs.center.x - 5) ? 1 : 0;
+            this->stateRight    |= (data->exp.classic.ljs.pos.x > data->exp.classic.ljs.center.x + 5) ? 1 : 0;
+            break;
+        default:
+            break;
```
Comment (Mefiresu) [inline RSDKv5/RSDK/Input/Wii/WiiInputDevice.cpp:62]: Consider moving `default` up to the `WPAD_EXP_NONE` case.

File: RSDKv5/RSDK/User/Core/UserCore.cpp [hunk 3]
```diff
@@ -500,7 +500,7 @@ void RSDK::LoadSettingsINI()
 
 #if !RETRO_USE_ORIGINAL_CODE
         customSettings.region                    = -1;
-        customSettings.confirmButtonFlip         = false;
+        customSettings.confirmButtonFlip         = true;
```
Comment (Mefiresu) [inline RSDKv5/RSDK/User/Core/UserCore.cpp:503]: Maybe keep the original value using `#if`.
```suggestion
#if RETRO_PLATFORM = RETRO_WII
        customSettings.confirmButtonFlip         = true;
#else
        customSettings.confirmButtonFlip         = false;
#endif
```

File: RSDKv5/RSDK/User/Core/UserCore.cpp [hunk 4]
```diff
@@ -614,7 +614,7 @@ void RSDK::SaveSettingsINI(bool32 writeToFile)
 #else
         WriteText(file, "language=%d\n", gameVerInfo.language);
 #endif
-
+	WriteText(file, "faceButtonFlip=%s\n", (customSettings.confirmButtonFlip ? "y" : "n"));
```
Comment (Mefiresu) [inline RSDKv5/RSDK/User/Core/UserCore.cpp:617]:
```suggestion
#if RETRO_PLATFORM == RETRO_WII
	WriteText(file, "faceButtonFlip=%s\n", (customSettings.confirmButtonFlip ? "y" : "n"));
#endif
```

**Review by Mefiresu — CHANGES_REQUESTED** [review]

File: RSDKv5/RSDK/Input/Wii/WiiInputDevice.cpp (hunk 2 above)
Comment (Anh302kay) [reply RSDKv5/RSDK/Input/Wii/WiiInputDevice.cpp]: `default` has been moved up to the `WPAD_EXP_NONE` case.

**Comment by Mefiresu** [comment]
Some feedback after testing mappings in general:
- The nunchuk analog's deadzone appears to be a bit low (might need to double check with other nunchuks) while classic's is fine
- Sideways Wiimote uses B to swap characters in Encore mode, which IMO is a bit hard to access in this configuration (swapping with the A button could be a better choice)

**DEV**: Now that there's effectively 3 different mappings, I feel like the variable names for button names can become confusing (RSDKv5 internally maps using Xbox-style controls with for example A to Confirm / B to Cancel). Maybe changing the variable names to fit their purpose would be clearer (so `stateA` could be `stateConfirm`, `stateB` could be `stateCancel`), or just writing comments in `WiiInputDevice.hpp` to explain their function in game.
Diving a bit deeper into it made me realize that the `confirmButtonFlip` setting is actually useless, and adds confusion. This actually caused me problems at first because the value wouldn't be written to `Settings.ini` if the file already existed.
It also appears that the Switch version actually has Confirm on B and Cancel on A (which physically corresponds to an Xbox controller layout). 
To be more accurate to how the Switch controllers are mapped, I suggest removing the `confirmButtonFlip` changes and manually swapping `stateA` with `stateB` in every mapping.

**Comment by Mefiresu** [comment]
Nevermind about the Switch layout, the buttons are indeed flipped.

IMG_20221001_201254

**Review by Mefiresu — APPROVED** [review]

**Pull request merged**